derive/src/utils.rs

```diff
@@ -37,32 +38,90 @@ where
 	})
 }
 
-/// Look for a `#[scale(index = $int)]` attribute on a variant. If no attribute
-/// is found, fall back to the discriminant or just the variant index.
-pub fn variant_index(v: &Variant, i: usize) -> TokenStream {
-	// first look for an attribute
-	let index = find_meta_item(v.attrs.iter(), |meta| {
-		if let NestedMeta::Meta(Meta::NameValue(ref nv)) = meta {
-			if nv.path.is_ident("index") {
-				if let Lit::Int(ref v) = nv.lit {
-					let byte = v
-						.base10_parse::<u8>()
-						.expect("Internal error, index attribute must have been checked");
-					return Some(byte);
+pub struct UsedIndexes {
```

Comment:
Missing docs.

Also this struct should only store the `used_set`. The function `from_iter` is not required.

You only need `variant_index`, which should be a verbatim copy of the old `variant_index` function plus that you add the resulting index to `used_set` and check if it already exists.

Reply:
added docs. 
we need current to assign incremental indexes to variants in ascending order also we need to traverse the variants first and collect all the assigned indexes before we can process them to check for duplicates and have a set of indexes already used, so that when we encounter a variant without explicit discriminant/`codec(index = ?)` we could assign an index using `UsedIndexes.current` by incrementing it until the `u8` is not contained inside the used set.